lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,259 @@
+
+    def _walk(path, top_level_dir, parent_dirs):
+        for item in os.listdir(path):
+            abs_path = os.path.join(path, item)
+            rel_base_dir = '' if path == top_level_dir else path[len(top_level_dir) + 1:]
+            rel_path = os.path.join(rel_base_dir, item)
+
+            if os.path.isdir(abs_path):
+                if item in ignore_dirs:
+                    display.vvv("Skipping %s for collection build" % abs_path)
+                    continue
+
+                if os.path.islink(abs_path):
+                    link_target = os.path.realpath(abs_path)
+
+                    if not link_target.startswith(top_level_dir):
+                        display.warning("Skipping %s as it is a symbolic link to a directory outside the collection"
+                                        % abs_path)
+                        continue
+
+                manifest_entry = entry_template.copy()
+                manifest_entry['name'] = rel_path
+                manifest_entry['ftype'] = 'dir'
+
+                manifest.append(manifest_entry)
+
+                _walk(abs_path, top_level_dir, parent_dirs)
+            else:
+                if item == 'galaxy.yml':
+                    continue
+                elif any([fnmatch.fnmatch(item, pattern) for pattern in ignore_files]):
+                    display.vvv("Skipping %s for collection build" % abs_path)
+                    continue
+
+                manifest_entry = entry_template.copy()
+                manifest_entry['name'] = rel_path
+                manifest_entry['ftype'] = 'file'
+                manifest_entry['chksum_type'] = 'sha256'
+                manifest_entry['chksum_sha256'] = secure_hash(abs_path, hash_func=sha256)
+
+                manifest.append(manifest_entry)
+
+    dir_stats = os.stat(collection_path)
+    parents = frozenset(((dir_stats.st_dev, dir_stats.st_ino),))
+    _walk(collection_path, collection_path, parents)
+
+    return manifest
+
+
+def _build_manifest(namespace, name, version, authors, tags, description, license, dependencies, **kwargs):
+
+    def _parse_author_info(author):
+        email = None
+        url = None
+
+        match = AUTHOR_REGEX.match(author)
+        if match:
+            author = match.group('name')
+            email = match.group('email')
+            url = match.group('url')
+
+        return {
+            'name': author,
+            'email': email,
+            'url': url
+        }
+    authors = [_parse_author_info(info) for info in authors]
+
+    dependencies = {name: {'version': version} for name, version in dependencies.items()}
+
+    manifest = {
+        'format': MANIFEST_FORMAT,
+        'name': "%s.%s" % (namespace, name),
+        'version': version,
+        'authors': authors,
+        'keywords': tags,
+        'description': description,
+        'license': license,
+        'license_file': None,  # TODO: thaumos to verify the need for license_file and readme
+        'readme': None,
```

Comment:
Is there a default directory the README should be placed in, the initial skeleton seems to be in collection dir but then others said it should be in the Docs dir. How is Mazer currently finding this file, is it expecting it to be somewhere in particular? Is there a preference for the default being .md or .rst?

Reply:
@jborean93 The  typical case is just a README.rst or README.md in the top level of the collection archive. At the moment, 'readme' is a required field though. https://github.com/ansible/galaxy/blob/devel/galaxy/importer/collection.py#L102

 But the galaxy code looks like it is preparing for 'readme' to be nullable, in which case it will default to
lREADME.rst' or 'README.md' in the top level dir. https://github.com/ansible/galaxy/blob/devel/galaxy/importer/utils/readme.py

Also note, that there is now a max filesize limit for the README of 512k (https://github.com/ansible/galaxy/blob/devel/galaxy/importer/utils/readme.py#L37). Though I don't think that needs to be enforce by 'build', since it will likely change for different galaxy servers.